feat: update prometheus agent daemonset proposal

slashexx opened the pull request:
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

Draft PR to start off the change of last years DaemonSet proposal.


## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [x] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification
<!-- How you tested it? How do you know it works? -->
Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

<!-- 
Your release note should be written in clear and straightforward sentences. Most often, users aren't familiar with
the technical details of your PR, so consider what they need to know when you write your release note.

Some brief examples of release notes:
- Add metadataConfig field to the Prometheus CRD for configuring how remote-write sends metadata information.
- Generate correct scraping configuration for Probes with empty or unset module parameter.
-->

```release-note
updated 2024's DaemonSet proposal to a complete version
```

mviswanathsai commented on Documentation/proposals/202405-agent-daemonset.md:
```diff
@@ -72,7 +72,52 @@ The current [PrometheusAgent CRD](https://prometheus-operator.dev/docs/platform/
+
+To prevent users from accidentally providing invalid configurations, we will implement validation rules using [Kubernetes' Common Expression Language (CEL)](https://kubernetes.io/docs/reference/using-api/cel/).
+
+These CEL rules will affect the DaemonSet mode and prevent:
```
(nit)
```suggestion
When `mode: Daemonset`, the following CEL rules will be applied:

Disallowed Configuration Options:
```

slashexx replied:
https://github.com/prometheus-operator/prometheus-operator/blob/9be68efafca772687346ffa8b8cc0cbeba88d1e7/pkg/apis/monitoring/v1alpha1/prometheusagent_types.go#L111

Done

mviswanathsai commented on Documentation/proposals/202405-agent-daemonset.md:
```diff
@@ -72,7 +72,52 @@ The current [PrometheusAgent CRD](https://prometheus-operator.dev/docs/platform/
+To prevent users from accidentally providing invalid configurations, we will implement validation rules using [Kubernetes' Common Expression Language (CEL)](https://kubernetes.io/docs/reference/using-api/cel/).
+
+These CEL rules will affect the DaemonSet mode and prevent:
+- Setting replicas 
```
```suggestion
- `replicas`
```

mviswanathsai commented on Documentation/proposals/202405-agent-daemonset.md:
```diff
@@ -72,7 +72,52 @@ The current [PrometheusAgent CRD](https://prometheus-operator.dev/docs/platform/
+These CEL rules will affect the DaemonSet mode and prevent:
+- Setting replicas 
+- Setting storage configuration 
+- Using ServiceMonitorSelector 
```
```suggestion
- `serviceMonitorSelector` 
```

mviswanathsai commented on Documentation/proposals/202405-agent-daemonset.md:
```diff
@@ -72,7 +72,52 @@ The current [PrometheusAgent CRD](https://prometheus-operator.dev/docs/platform/
+- Setting replicas 
+- Setting storage configuration 
+- Using ServiceMonitorSelector 
+- Setting shards 
```
```suggestion
- `shards`
```

mviswanathsai commented on Documentation/proposals/202405-agent-daemonset.md:
```diff
@@ -72,7 +72,52 @@ The current [PrometheusAgent CRD](https://prometheus-operator.dev/docs/platform/
+- Using ServiceMonitorSelector 
+- Setting shards 
+- Setting persistentVolumeClaimRetentionPolicy 
+- Using serviceMonitorNamespaceSelector 
```
```suggestion
- `serviceMonitorNamespaceSelector` 
```

mviswanathsai commented on Documentation/proposals/202405-agent-daemonset.md:
```diff
@@ -72,7 +72,52 @@ The current [PrometheusAgent CRD](https://prometheus-operator.dev/docs/platform/
+- Setting shards 
+- Setting persistentVolumeClaimRetentionPolicy 
+- Using serviceMonitorNamespaceSelector 
+- Setting serviceName 
```
hmm... I am not sure if this is unacceptable for a Daemonset. I think it is a valid use case to have a governing service that exposes all the Daemonsets.
CC: @simonpasquier @slashpai

mviswanathsai commented on Documentation/proposals/202405-agent-daemonset.md:
```diff
@@ -72,7 +72,52 @@ The current [PrometheusAgent CRD](https://prometheus-operator.dev/docs/platform/
+CEL validation will provide immediate feedback during `kubectl apply` but we will need runtime validation logic in the controller as a fallback mechanism. This fallback will be integrated directly in the `PrometheusAgent` reconciler loop.
+
+This is mainly because :
+1. CEL validation will require Kubernetes version 1.25+ and hence not all users might have CEL supported clusters.
```
This is a valid concern. But Kubernetes is currently at major release `v1.33.0`. Would it be fine to not support this as well? Thoughts? @simonpasquier @slashpai @nicolastakashi

mviswanathsai reviewed (commented):
Great start! :partying_face: 
I have a few concerns that I have mentioned.

slashexx commented on Documentation/proposals/202405-agent-daemonset.md:
```diff
@@ -72,7 +72,52 @@ The current [PrometheusAgent CRD](https://prometheus-operator.dev/docs/platform/
+- Setting serviceName 
+
+Additionally, it enforces the following rules:
+- Require PodMonitorSelector in DaemonSet mode
```
Yup looked into this and it's optional so I've removed it now.

slashexx commented:
@mviswanathsai I've added updated text about `ServiceMonitors` and `EndpointSlices`, requesting review !

slashpai commented on Documentation/proposals/202405-agent-daemonset.md:
```diff
@@ -46,6 +47,15 @@ In specific, the MVP will need to:
+* **EndpointSlice Integration**: Implement support for `serviceDiscoveryRole: EndpointSlice` in PrometheusAgent CRD to use EndpointSlice API instead of classic Endpoints for service discovery.
+* **Performance Optimization**: Reduce Kubernetes API server load by using the more efficient and scalable EndpointSlice API, which distributes endpoint information across multiple smaller objects instead of one large Endpoints object.
+* **DaemonSet Compatibility**: Ensure EndpointSlice discovery works correctly with DaemonSet mode for node-local service monitoring.
+* **Validation and Fallback**: Implement proper validation to ensure EndpointSlice is only used when supported by the Kubernetes cluster (v1.21+), with graceful fallback to classic Endpoints on older clusters.
```
I am wondering whether we need to provide fallback as 1.21 itself is way too older version
@simonpasquier wdyt?

slashpai commented:
As discussed in the call we can explore the servicemonitor option as a separate feature gate so it doesn't block making current feature GA.

slashpai commented on Documentation/proposals/accepted/202405-agent-daemonset.md:
```diff
@@ -84,6 +85,61 @@ The current [PrometheusAgent CRD](https://prometheus-operator.dev/docs/platform/
+- `storage`
+- `shards`
+- `persistentVolumeClaimRetentionPolicy`
+- `scrapeConfigSelector`
```
all other selectors which are invalid in daemonset mode too here

slashpai commented on Documentation/proposals/accepted/202405-agent-daemonset.md:
```diff
@@ -84,6 +85,61 @@ The current [PrometheusAgent CRD](https://prometheus-operator.dev/docs/platform/
+  - rule: "self.mode == 'DaemonSet' ? !has(self.persistentVolumeClaimRetentionPolicy) : true"
+    message: "persistentVolumeClaimRetentionPolicy field is not allowed when mode is 'DaemonSet'"
+  - rule: "!(has(self.mode) && self.mode == 'DaemonSet' && has(self.scrapeConfigSelector))"
+    message: "scrapeConfigSelector cannot be set when mode is DaemonSet"
```
looks like some more missing here from whats been implemented, mainly the selectors

slashpai commented on Documentation/proposals/accepted/202405-agent-daemonset.md:
```diff
@@ -84,6 +85,61 @@ The current [PrometheusAgent CRD](https://prometheus-operator.dev/docs/platform/
+	}
+	if spec.ScrapeConfigSelector != nil {
+		return fmt.Errorf("cannot configure scrapeConfigSelector when using DaemonSet mode")
+	}
```
nit: other selectors too

slashpai commented on Documentation/proposals/accepted/202405-agent-daemonset.md:
```diff
@@ -84,6 +85,97 @@ The current [PrometheusAgent CRD](https://prometheus-operator.dev/docs/platform/
+- `scrapeConfigNamespaceSelector`
+- `probeSelector`
+- `probeNamespaceSelector`
+- `serviceMonitorSelector`
```
since we have added a secondary goal of exploring servicemonitor with endpointslice I am a bit confused about adding this here. Or should that be a different proposal altogether to avoid confusion?

any thoughts?

simonpasquier replied:
maybe add a general note after the list that: `some validations may be lifted in the future (for instance if the DaemonSet mode supports ServiceMonitor resources)`?

simonpasquier commented on Documentation/proposals/accepted/202405-agent-daemonset.md:
```diff
@@ -84,6 +85,97 @@ The current [PrometheusAgent CRD](https://prometheus-operator.dev/docs/platform/
 
+#### 6.1.1 CEL Validation rules
+
+When `mode:DaemonSet`, the following CEL rules will be applied:
```
```suggestion
When `mode:DaemonSet`, the following CEL rules will be applied to prevent access to these fields:
```

simonpasquier commented on Documentation/proposals/accepted/202405-agent-daemonset.md:
```diff
@@ -84,6 +85,97 @@ The current [PrometheusAgent CRD](https://prometheus-operator.dev/docs/platform/
+CEL validation will provide immediate feedback during `kubectl apply` but we will need runtime validation logic in the controller as a fallback mechanism. This fallback will be integrated directly in the `PrometheusAgent` reconciler loop.
+
+This is mainly because :
+1. CEL validation will require Kubernetes version 1.25+ and hence not all users might have CEL supported clusters.
```
CEL has been there for quite some time now (1.25 is 3 years-old) so we can also update our min requirements (no cloud provider should be offering 1.25 by now).

slashexx replied:
@simonpasquier then should we drop the runtime validations entirely ?

simonpasquier commented on Documentation/proposals/accepted/202405-agent-daemonset.md:
```diff
@@ -115,6 +207,66 @@ We'll go with this option, because it filters targets right at discovery time, a
+  kubernetes_sd_configs:
+  - role: endpointslice  # Instead of "endpoints"
+    namespaces:
+      names: [default]
```
how does it ensure that only targets from the node are selected?

slashexx replied:
Since we'll explore endpointslices later on, I think I should drop these fields for now and add them when the non-goal is completed, what do you think @slashpai @mviswanathsai @simonpasquier ?

slashexx commented:
cc @slashpai @mviswanathsai ptal at the comments once and let's finalize.

slashexx commented:
gentle ping @slashpai @mviswanathsai

slashpai reviewed (dismissed):
lgtm

@simonpasquier if no objections we can merge this :)

slashexx committed:
docs: add CEL validation rules to agent daemonset proposal

Add comprehensive CEL (Common Expression Language) validation rules
to the agent daemonset proposal documentation, including:
- x-kubernetes-validations for field validation
- Runtime fallback logic for validation handling
- Service monitor section refactoring
- Endpointslice section documentation
- Related field blocks and validation requirements

slashpai commented:
@simonpasquier could you please take a look again

slashexx commented:
@slashpai considering this PR mentions the runtime validations in  #7602 , we should finalize on that first, once it is done, I suppose moving this proposal to implemented should be in order ?

simonpasquier reviewed (approved):
lgtm
